packages/core/src/components/tabs2/tabs.tsx

```diff
@@ -0,0 +1,293 @@
+ * Licensed under the BSD-3 License as modified (the “License”); you may obtain a copy
+ * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+ * and https://github.com/palantir/blueprint/blob/master/PATENTS
+ */
+
+import * as classNames from "classnames";
+import * as PureRender from "pure-render-decorator";
+import * as React from "react";
+
+import { AbstractComponent } from "../../common/abstractComponent";
+import * as Classes from "../../common/classes";
+import * as Keys from "../../common/keys";
+import { IProps } from "../../common/props";
+import { safeInvoke } from "../../common/utils";
+
+import { ITab2Props, Tab2, TabId } from "./tab";
+import { generateTabPanelId, generateTabTitleId, TabTitle } from "./tabTitle";
+
+export const Expander: React.SFC<{}> = () => <div className="pt-flex-expander" />;
+
+type TabElement = React.ReactElement<ITab2Props & { children: React.ReactNode }>;
+
+const TAB_SELECTOR = `.${Classes.TAB}`;
+
+export interface ITabs2Props extends IProps {
+    /**
+     * Whether the selected tab indicator should animate its movement.
+     * @default true
+     */
+    animate?: boolean;
+
+    /**
+     * Initial selected tab `id`, for uncontrolled usage.
+     * Note that this prop refers only to `<Tab>` children; other types of elements are ignored.
+     * @default first tab
+     */
+    defaultSelectedTabId?: TabId;
+
+    /**
+     * Unique identifier for this `Tabs` container. This will be combined with the `id` of each
+     * `Tab` child to generate ARIA accessibility attributes. IDs are required and should be
+     * unique on the page to support server-side rendering.
+     */
+    id: TabId;
+
+    /**
+     * Whether inactive tab panels should be removed from the DOM and unmounted in React.
+     * This can be a helpful performance enhancement but requires careful support for
+     * unmounting and remounting properly.
+     * @default false
+     */
+    renderActiveTabPanelOnly?: boolean;
+
+    /**
+     * Selected tab `id`, for controlled usage.
+     * Providing this prop will put the component in controlled mode.
+     * Unknown ids will result in empty selection (no errors).
+     */
+    selectedTabId?: TabId;
+
+    /**
+     * Whether to show tabs stacked vertically on the left side.
+     * @default false
+     */
+    vertical?: boolean;
+
+    /**
+     * A callback function that is invoked when a tab in the tab list is clicked.
+     */
+    onChange?(newTabId: TabId, prevTabId: TabId): void;
+}
+
+export interface ITabs2State {
+    indicatorWrapperStyle?: React.CSSProperties;
+    selectedTabId?: TabId;
+}
+
+@PureRender
+export class Tabs2 extends AbstractComponent<ITabs2Props, ITabs2State> {
+    /** Insert a `Tabs2.Expander` between any two children to right-align all subsequent children. */
+    public static Expander = Expander;
+
+    public static Tab = Tab2;
+
+    public static defaultProps: ITabs2Props = {
+        animate: true,
+        id: "",
```

Comment:
I realized recently that we can start using partial types for default props and retain type safety for consumers. Sometimes we only want to supply defaults for a few props but not make those fields we omit optional in the props interface. I believe this should work for us:

```ts
public static defaultProps: Partial<ITabs2Props> = {
    animate: true,
    renderActiveTabPanelOnly: false,
    vertical: false,
};
```